[lldb] Adding pipe support to lldb_private::MainLoopWindows.

ashgti opened the pull request:
This updates MainLoopWindows to support events for reading from a pipe (both anonymous and named pipes) as well as sockets.

This unifies both handle types using `WSAWaitForMultipleEvents` which can listen to both sockets and handles for change events.

This should allow us to unify how we handle watching pipes/sockets on Windows and Posix systems.

ashgti committed:
[lldb] Adding file and pipe support to lldb_private::MainLoopWindows.

This updates MainLoopWindows to support events for reading from a file and a socket type.

This unifies both handle types using WaitForMultipleEvents which can listen to both sockets and files for change events.

This should allow us to unify how we handle watching files/pipes/sockets on Windows and Posix systems.

labath commented on lldb/source/Host/windows/MainLoopWindows.cpp:
```diff
@@ -44,36 +37,61 @@ MainLoopWindows::~MainLoopWindows() {
-      WSAWaitForMultipleEvents(events.size(), events.data(), FALSE,
-                               ToTimeout(GetNextWakeupTime()), FALSE);
+  while (true) {
+    DWORD timeout = INFINITY;
+    std::optional<lldb_private::MainLoopBase::TimePoint> deadline =
+        GetNextWakeupTime();
+    if (deadline) {
+      // Check how much time is remaining, we may have woken up early for an
+      // unrelated reason on a file descriptor (e.g. a stat was triggered).
+      std::chrono::milliseconds remaining =
+          std::chrono::duration_cast<std::chrono::milliseconds>(
+              deadline.value() - std::chrono::steady_clock::now());
+      if (remaining.count() <= 0)
+        return events.size() - 1;
+      timeout = remaining.count();
+    }
```
Is this any different from
```
DWARF timeout = ToTimeout(GetNextWakeupTime());
if (timeout == 0)
  return events.size() - 1;
```
(using the original definition of the ToTimeout function) ?

ashgti replied:
Reverted this and some of the other changes to MainLoopWindows::Poll, its more like the previous version but we now have additional helpers for other types.

labath commented on lldb/source/Host/windows/MainLoopWindows.cpp:
```diff
@@ -44,36 +37,61 @@ MainLoopWindows::~MainLoopWindows() {
+            ->second.object_sp->HasReadableData())
+      return result - WAIT_OBJECT_0;
+
+    // If no handles are actually ready then yield the thread to allow the CPU
+    // to progress.
+    std::this_thread::yield();
```
What exactly are the situations when we reach this? I don't mind spinning here once or twice, but I'd hate to see this turn into busy wait...

ashgti replied:
WaitForSingleObject/WaitForMultipleObjects will signal on any change to a file handle, so we can't really differentiate the specific event that caused the change. I've updated the code to use a zero byte read instead and now we shouldn't need to keep trying.

labath commented on lldb/source/Host/common/File.cpp:
```diff
@@ -274,7 +276,23 @@ int NativeFile::GetDescriptor() const {
+bool NativeFile::HasReadableData() {
+#ifdef _WIN32
+  DWORD available_bytes = 0;
+  return !PeekNamedPipe((HANDLE)_get_osfhandle(GetDescriptor()), NULL, 0, NULL,
```
This will only work for "files" that are actually pipes, right?

If so, I don't think that's a good long-term direction. My medium-to-long term plan would be to make Pipe an IOObject, but that requires splitting it into two (one IOObject for each end). I don't think that would be a particularly hard refactor, but if you don't want to take that on, maybe we can find a way to pass a Pipe to MainLoop directly (and keep NativeFile out of it)? I'd be fine with an RegisterReadObject overload that takes a Pipe (and ignores the write end of the pipe -- we don't  select on writes anyway).

ashgti replied:
I've split this up for checking between files and pipes. For files, we'll get notified anytime the file changes on disk, for pipes we'll use the zero byte read in a monitor thread to signal when new data is available. That strategy should work with both named pipes and anonymous pipes. I'm checking the file type we get out of the IOObject though, so this should work if you passed in `stdin` as well.

labath commented on lldb/source/Host/common/Socket.cpp:
```diff
@@ -313,8 +318,39 @@ Socket::DecodeHostAndPort(llvm::StringRef host_and_port) {
+    m_waitable_handle = WSACreateEvent();
+    assert(m_waitable_handle != WSA_INVALID_EVENT);
+    if (WSAEventSelect(m_socket, m_waitable_handle,
+                       FD_ACCEPT | FD_READ | FD_WRITE) != 0)
```
Does this mean that the event will be triggered whenever the socket is writable (triggering the busy wait I mentioned before)?  That would definitely be undesirable. 

If WSAEventSelect supports it, we might be able to deal with that by using separate events for reads and writes. The question is somewhat academic because noone is currently selecting for writes, but it's interesting from the perspective of where should that event live. You're moving it from the MainLoop to the Socket, which kind of makes sense, but it may not be possible if there isn't a single event (or a set of events) that can be used for everything.

In that world, I think that the original setup (where the event is managed inside the main loop) makes more sense as the loop can `WSAEventSelect` precisely the events it is interested it. And it should mostly (as long as they don't attempt that in parallel) work with sockets that are added to multiple main loops (which definitely works on posix).

In principle, I'd be fine with a one-mainloop-per-socket restriction, but we'd need to do some refactoring to make that happen as we currently in fact use multiple instances of selector in some situations (ConnectionFileDescriptor uses a SelectHelper internally and lldb-server adds the connection to its top level main loop).

labath reviewed (commented):
Thank you again for looking into this. 

I think that the most important question we need to figure out is what's the danger of this turning into a busy wait. A function like `HasReadableData` could be useful in some situation, so I don't have a problem with it per-se, but the way you're using it makes me very suspicious that can turn into a busy wait under some circumstances. I think we at least need to know what those circumstances are, so we can evaluate whether that worth the (very tangible) benefit of being able to use this functionality portably.

labath commented:
In case it helps, [here's](https://github.com/llvm/llvm-project/commit/6675e03763c7d64e3335f1582c0fd7bc639be7d7) a method to wait on a pipe using zero-length reads (inspired by the [libdispatch](https://github.com/swiftlang/swift-corelibs-libdispatch/blob/fdeda69b00a5b6de8be568c908f3ea2b89c37065/src/event/event_windows.c#L278) implementation you found earlier). In case waiting on the pipe handle itself produces spurious wakeups, I think this should be a fairly reliable alternative. The actual implementation probably should handle the zero-length write like libdispatch, but retrying there should be fine as it should only happen if/when someone does a zero-length write. I'm hoping we can ignore PIPE_NOWAIT pipes since they're kind of deprecated.

ashgti committed:
Adopting a different strategy for monitoring pipes and files.

For pipes, we use a monitor thread that performs a zero byte read to detect changes before signaling.

For files, WaitFor{SingleObject,MultipleObjects}() is signaled if the file changes on disk.

labath reviewed (commented):
This looks much better. My two main questions now are:
- do we need regular file support
- do we need the pipe *thread*

(I think/hope the answer is "no" on both counts).

labath commented:
> > * do we need regular file support
> 
> We don't strictly need this for anything I am aware of, I can remove it for now. A 'better' solution would be to stat the file and then watch the directory for changes and re-stat the file. Thats used in libuv (nodejs). I think libdispatch on Windows doesn't actually do anything for files, so you don't get notified if a file on disk changes.

Yeah, let's remove that. I think that "watching for file/directory" changes is a completely different feature..

> 
> > * do we need the pipe _thread_
> 
> In order to support both anonymous pipes, like stdin, and named pipes we would need the thread.
> 
> We could specialize this further and have the thread only for anonymous pipes and use a different strategy for named pipes, but the current solution works for both.

I see. I take it that's because anonymous pipes are created without FILE_FLAG_OVERLAPPED. Btw, my understanding is that an anonymous pipe is just a named pipe with a funny name (at least until windows 10, or something) and some with some hardcoded flags (like the absence of overlapped I/O). This is why lldb creates "anonymous pipes" as named -- so it can enable overlapped I/O. So for internal pipe uses, I think we're safe. It gets trickier for pipes that get passed from the outside, but  I wouldn't be surprised if VSCode did something similar -- it's likely it also wants to do some overlapped I/O.

Could you check if that is the case?

I'm not sure what that would mean for the implementation (I like the generality, but I also hate the idea of creating a thread for every pipe object), but I think it'd be good to have know this.

ashgti committed:
Removing disk file support and reverting unrelated changes.

ashgti committed:
Apply clang-format.

ashgti commented on lldb/source/Host/windows/MainLoopWindows.cpp:
```diff
@@ -21,7 +24,9 @@
 using namespace lldb_private;
 
-static DWORD ToTimeout(std::optional<MainLoopWindows::TimePoint> point) {
+namespace {
```
Done.

ashgti committed:
Apply clang-format again.

ashgti commented:
On my windows VM, all tests are passing now and it looks like this is also working for Linux. I think the current CI check is a no-op on Windows for lldb.

labath commented on lldb/source/Host/windows/MainLoopWindows.cpp:
```diff
@@ -31,6 +34,122 @@ static DWORD ToTimeout(std::optional<MainLoopWindows::TimePoint> point) {
+  }
+
+  void WillPoll() override {
+    if (!m_monitor_thread.joinable())
```
What would you say to creating the thread in the constructor? It would remove the `joinable()` calls, and thing that adding a pipe to the loop is quickly followed by a poll operation anyway.

labath commented on lldb/source/Host/windows/MainLoopWindows.cpp:
```diff
@@ -31,6 +34,122 @@ static DWORD ToTimeout(std::optional<MainLoopWindows::TimePoint> point) {
+            /*bInitialState=*/FALSE, NULL)),
+        m_handle(handle), m_ready(CreateEventW(NULL, /*bManualReset=*/FALSE,
+                                               /*bInitialState=*/FALSE, NULL)) {
+    assert(event && ready);
```
```suggestion
    assert(m_event && m_ready);
```

labath commented on lldb/source/Host/windows/MainLoopWindows.cpp:
```diff
@@ -44,25 +163,20 @@ MainLoopWindows::~MainLoopWindows() {
-    int result = WSAEventSelect(fd.first, WSA_INVALID_EVENT, 0);
-    assert(result == 0);
-    UNUSED_IF_ASSERT_DISABLED(result);
+  for (auto &[_, fd_info] : m_read_fds) {
+    fd_info.event->DidPoll();
   }
```
```suggestion
  for (auto &[_, fd_info] : m_read_fds)
    fd_info.event->DidPoll();
```

labath commented on lldb/unittests/Host/MainLoopTest.cpp:
```diff
@@ -79,7 +79,68 @@ TEST_F(MainLoopTest, ReadObject) {
+
+  char X = 'X';
+  size_t len = sizeof(X);
+  ASSERT_THAT_EXPECTED(pipe.Write(&X, len), llvm::Succeeded());
```
```suggestion
  ASSERT_THAT_EXPECTED(pipe.Write(&X, len), llvm::HasValue(1));
```

labath commented on lldb/unittests/Host/MainLoopTest.cpp:
```diff
@@ -79,7 +79,68 @@ TEST_F(MainLoopTest, ReadObject) {
+          // pipe is empty, and we should not be called anymore.
+          char X;
+          size_t len = sizeof(X);
+          EXPECT_THAT_ERROR(r->Read(&X, len).ToError(), llvm::Succeeded());
+          EXPECT_EQ(len, sizeof(X));
```
```suggestion
          EXPECT_THAT_EXPECTED(r->Read(&X, len).ToError(), llvm::HasValue(1));
```

labath reviewed (approved):
> I made a sample binary (https://gist.github.com/ashgti/1649fde8ef28783ace1e414f87a6b680) and it seems stdin is opened as an anonymous pipe.

Got it. Thanks for checking it out. I guess we have no choice, but to create the extra thread. It's unfortunate that windows doesn't allow you to do async I/O if the file handle wasn't created as such. The "just create it the right way from the start" argument makes sense when everything happens within a single process, but this use case (passing a pipe handle from one process to another) is kind of a counterexample to that.

I also wouldn't be surprised if we needed to add threads to wait for other kinds of events in the future. One thing that would be occasionally useful to do is to wait for some data to arrive (e.g. a process writes something to a pipe) *OR* for a process to die (e.g. the writing process crashes). Linux can pull that off without threads using a pidfd, but I'm not sure all systems can.

It's also reassuring to know that we can avoid the extra thread if we need to for pipes we have create ourselves, but I think we don't need to do that now. I think most of the internal uses of pipes would be better off by switching to `AddPendingCallback` (which internally uses a pipe on posix, but directly signals an event on windows).

ashgti committed:
Apply suggestions from code review

Co-authored-by: Pavel Labath <pavel@labath.sk>

ashgti committed:
Update lldb/source/Host/windows/MainLoopWindows.cpp

Co-authored-by: Pavel Labath <pavel@labath.sk>

ashgti committed:
Fixing an assert.

ashgti commented:
I was running my tests in `Release` mode, so I missed that one of the `assert` calls had an invalid variable name. Fixed in https://github.com/llvm/llvm-project/commit/a79404e50cf5308bb76cc4eaf36cb8eaea9ab66a

mstorsjo commented:
This broke compilation on mingw (for 64 bit targets, a 32 bit test build is still pending):
```
../../lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp:455:29: error: cast from pointer to smaller type 'lldb::socket_t' (aka 'unsigned int') loses information
  455 |     select_helper.FDSetRead((lldb::socket_t)handle);
      |                             ^~~~~~~~~~~~~~~~~~~~~~
../../lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp:497:39: error: cast from pointer to smaller type 'lldb::socket_t' (aka 'unsigned int') loses information
  497 |         if (select_helper.FDIsSetRead((lldb::socket_t)handle))
      |                                       ^~~~~~~~~~~~~~~~~~~~~~
2 warnings and 2 errors generated.
```

mstorsjo commented:
> They put up #146632 to fix it, if one of you wants to check it works.

Thanks - it looks like that does fix some issues, but it only lets the build proceed slightly further to hit other errors caused by this patch:
```
../../lldb/source/Host/windows/MainLoopWindows.cpp:120:3: error: no type named 'thread' in namespace 'std'; did you mean 'Thread'?
  120 |   std::thread m_monitor_thread;
      |   ^~~~~~~~~~~
      |   Thread
../../lldb/include/lldb/lldb-forward.h:241:7: note: 'Thread' declared here
  241 | class Thread;
      |       ^
../../lldb/source/Host/windows/MainLoopWindows.cpp:67:31: error: no member named
 'thread' in namespace 'std'
   67 |       m_monitor_thread = std::thread(&PipeEvent::Monitor, this);
      |                               ^~~~~~

../../lldb/source/Host/windows/MainLoopWindows.cpp:213:33: error: no viable overloaded '='
  213 |     m_read_fds[waitable_handle] = {
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~
  214 |         std::make_unique<PipeEvent>((HANDLE)waitable_handle), callback};
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../lldb/include/lldb/Host/windows/MainLoopWindows.h:56:10: note: candidate function (the implicit copy assignment operator) not viable: cannot convert initializer list argument to 'const FdInfo'
   56 |   struct FdInfo {
      |          ^~~~~~
../../lldb/include/lldb/Host/windows/MainLoopWindows.h:56:10: note: candidate function (the implicit move assignment operator) not viable: cannot convert initializer list argument to 'FdInfo'
   56 |   struct FdInfo {
      |          ^~~~~~
```

That seems to be easy to fix with a plain `#include <thread>` in `MainLoopWindows.cpp` though.

ashgti commented:
I'll update my PR to add `#include <thread>`

slydiman commented:
Please look at [lldb-remote-linux-win](https://lab.llvm.org/buildbot/#/builders/197/builds/6610)
Probably this test is flaky now.
```
FAIL: lldb-unit::HostTests.exe/MainLoopTest/NoSpuriousPipeReads
[view all 27 lines](https://lab.llvm.org/buildbot/#/builders/197/builds/6610/steps/16/logs/FAIL__lldb-unit__HostTests_exe_MainLoopTest_NoSpur) 
******************** TEST 'lldb-unit :: Host/./HostTests.exe/31/87' FAILED ********************
Script(shard):
--
GTEST_OUTPUT=json:C:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\unittests\Host\.\HostTests.exe-lldb-unit-11488-31-87.json GTEST_SHUFFLE=0 GTEST_TOTAL_SHARDS=87 GTEST_SHARD_INDEX=31 C:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\unittests\Host\.\HostTests.exe
--
Script:
--
C:\buildbot\as-builder-10\lldb-x-aarch64\build\tools\lldb\unittests\Host\.\HostTests.exe --gtest_filter=MainLoopTest.NoSpuriousPipeReads
--
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\unittests\Host\MainLoopTest.cpp(141): error: Expected equality of these values:
  1u
    Which is: 1
  callback_count
    Which is: 2
C:\buildbot\as-builder-10\lldb-x-aarch64\llvm-project\lldb\unittests\Host\MainLoopTest.cpp:141
Expected equality of these values:
  1u
    Which is: 1
  callback_count
    Which is: 2
********************
```

slydiman commented:
The test lldb-unit::HostTests.exe/MainLoopTest/NoSpuriousPipeReads is flaky on Windows.
Please fix it.

https://lab.llvm.org/buildbot/#/builders/211/builds/231
https://lab.llvm.org/buildbot/#/builders/197/builds/6659
